src/libostree/ostree-repo-private.h

```diff
@@ -126,6 +129,24 @@ typedef struct {
   char checksum[OSTREE_SHA256_STRING_LEN+1];
 } OstreeDevIno;
 
+/* A MemoryCacheRef is an in-memory cache of objects (currently just DIRMETA).  This can
+ * be used when performing an operation that traverses a repository in someway.  Currently,
+ * the primary use case is ostree_repo_checkout_at() avoiding lots of duplicate dirmeta
+ * lookups.
+ */
+typedef struct {
+  OstreeRepo *repo;
+} OstreeRepoMemoryCacheRef;
```

Comment:
Right; I think `OstreeRepo` should boil down mostly to a directory fd.  Right now e.g. rpm-ostree will potentially hold on to the sysroot for a while (which I'd like to have hold a repo), and if we didn't do this, the cache would grow without bounds.

Including it in the repo itself is *partly* about avoiding argument plumbing - right now it'd be hugely painful since we'd have to rewrite `OstreeRepoFile`.  But it's also about the fact that in the future we want to do concurrent/mult-threaded checkout for example, and the repo acts as a shared cache there.